lib/cog/command/service/memory.ex

```diff
@@ -0,0 +1,168 @@
+  is successfully joined. If the key exists and is a list and the value is a
+  list, the value is joined to the end of the list. If the key does not exist,
+  the value is joined to an empty list. If either value is not a list `{:error,
+  :value_not_list}` is returned.
+  """
+  def join(token, key, value) when is_list(value),
+    do: GenServer.call(__MODULE__, {:join, token, key, value})
+  def join(_token, _key, _value),
+    do: {:error, :value_not_list}
+
+  @doc """
+  Replaces or sets the given key with the value. Returns `{:ok, value}`.
+  """
+  def replace(token, key, value),
+    do: GenServer.call(__MODULE__, {:replace, token, key, value})
+
+  @doc """
+  Deletes the given key. Returns `{:ok, deleted_value}` when successfully
+  deleted or `{:error, :unknown_key}` if it doesn't exist.
+  """
+  def delete(token, key),
+    do: GenServer.call(__MODULE__, {:delete, token, key})
+
+  def init([tid]) do
+    Logger.info("Starting with token table #{inspect tid}")
+    {:ok, %__MODULE__{tid: tid}}
+  end
+
+  def handle_call({:fetch, token, key}, _from, %__MODULE__{tid: tid} = state) do
+    result = ets_lookup(tid, {token, key})
+    {:reply, result, state}
+  end
+
+  def handle_call({:accum, token, key, value}, _from, %__MODULE__{tid: tid} = state) do
+    monitor_executor(token, tid)
+
+    result = with {:ok, existing_value}    <- ets_lookup_accum_list(tid, {token, key}),
+                  {:ok, accumulated_value} <- ets_insert(tid, {token, key}, existing_value ++ [value]),
+                  do: {:ok, accumulated_value}
+
+    {:reply, result, state}
+  end
+
+  def handle_call({:join, token, key, value}, _from, %__MODULE__{tid: tid} = state) when is_list(value) do
+    monitor_executor(token, tid)
+
+    result = with {:ok, existing_value}    <- ets_lookup_join_list(tid, {token, key}),
+                  {:ok, accumulated_value} <- ets_insert(tid, {token, key}, existing_value ++ value),
+                  do: {:ok, accumulated_value}
+
+    {:reply, result, state}
+  end
+
+  def handle_call({:replace, token, key, value}, _from, %__MODULE__{tid: tid} = state) do
+    monitor_executor(token, tid)
+
+    result = ets_insert(tid, {token, key}, value)
+    {:reply, result, state}
+  end
+
+  def handle_call({:delete, token, key}, _from, %__MODULE__{tid: tid} = state) do
+    result = ets_delete(tid, {token, key})
+    {:reply, result, state}
+  end
+
+  def handle_info({:DOWN, monitor_ref, :process, pid, reason}, %__MODULE__{tid: tid} = state) do 
+    Logger.debug("Process #{inspect pid} went down (#{inspect reason}); removing its memory storage")
+
+    case :ets.lookup(tid, monitor_ref) do
+      [{^monitor_ref, token}] ->
+        :ets.match_delete(tid, {{token, :_}, :_})
+        :ets.delete(tid, monitor_ref)
+      [] ->
+        Logger.warn("Unknown monitor ref #{inspect monitor_ref} for pid #{inspect pid} going down for #{inspect reason}")
+    end
+
+    {:noreply, state}
+  end
+
+  defp monitor_executor(token, tid) do
```

Comment:
I think we're going to stack up monitors with this code, one for each operation that gets called in the pipeline. We should probably just monitor the process if we're not already monitoring it.